File: app/src/main/java/com/icerrate/popularmovies/view/common/BaseActivity.java
package com.icerrate.popularmovies.view.common;

import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import com.icerrate.popularmovies.R;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * @author Ivan Cerrate.
 */

public abstract class BaseActivity extends AppCompatActivity implements BaseFragmentListener {

    private Unbinder unbinder;

    protected Toolbar toolbar;

    @Override
    public void setContentView(@LayoutRes int layoutResID) {
        super.setContentView(layoutResID);
        unbinder = ButterKnife.bind(this);
        toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        saveInstanceState(outState);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(R.anim.none, R.anim.fade_out);
    }

    @Override
    protected void onDestroy() {
        unbinder.unbind();
        super.onDestroy();
    }

    @Override
    public void setNavigationToolbar(boolean navigation) {
        if (navigation && getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }

    @Override
    public void replaceFragment(int containerId, Fragment fragment) {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.add(containerId, fragment);
        transaction.commit();
    }

    @Override
    public void setTitle(String title) {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(title);
        }
    }

    protected void saveInstanceState(Bundle outState) {
        //Needs to be empty
    }

    protected void restoreInstanceState(Bundle savedState) {
        //Needs to be empty
    }
}

File: app/src/main/java/com/icerrate/popularmovies/view/common/BaseActivity.kt
package com.icerrate.popularmovies.view.common

import android.os.Bundle
import android.view.LayoutInflater
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentTransaction
import androidx.appcompat.app.AppCompatActivity
import android.view.MenuItem
import androidx.viewbinding.ViewBinding
import com.icerrate.popularmovies.R

/**
 * @author Ivan Cerrate.
 */

abstract class BaseActivity<V: ViewBinding> (
    private val inflate: (LayoutInflater) -> V
) : AppCompatActivity(), BaseFragmentListener {

    protected lateinit var binding: V

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = inflate(layoutInflater)
        setContentView(binding.root)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressedDispatcher.onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        saveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    override fun finish() {
        super.finish()
        overridePendingTransition(R.anim.none, R.anim.fade_out)
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    override fun setNavigationToolbar(navigation: Boolean) {
        if (navigation && supportActionBar != null) {
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }
    }

    override fun replaceFragment(containerId: Int, fragment: Fragment) {
        val transaction: FragmentTransaction = supportFragmentManager.beginTransaction()
        transaction.add(containerId, fragment)
        transaction.commit()
    }

    protected open fun saveInstanceState(outState: Bundle) {
        // Needs to be empty
    }

    protected open fun restoreInstanceState(savedState: Bundle) {
        // Needs to be empty
    }
}